File: app/src/main/java/com/example/chap18example/MainActivity.java
package com.example.chap18example;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    EditText userName, userMessage;
    Button counter;
    CheckBox remember;

    String name, message;
    int count = 0;
    boolean isChecked;

    SharedPreferences sharedPreferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        userName = findViewById(R.id.editTextText);
        userMessage = findViewById(R.id.editTextTextMultiLine);
        counter = findViewById(R.id.button);
        remember = findViewById(R.id.checkBox);

        counter.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                count++;
                counter.setText("" + count);
            }
        });

        retrieveData();
    }

    @Override
    protected void onPause() {
        super.onPause();
        saveData();
    }

    public void saveData() {
        sharedPreferences = getSharedPreferences("saveData", Context.MODE_PRIVATE);
        name = userName.getText().toString();
        message = userMessage.getText().toString();
        isChecked = remember.isChecked();

        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString("name", name);
        editor.putString("message", message);
        editor.putInt("count", count);
        editor.putBoolean("isChecked", isChecked);

        editor.commit();
        Toast.makeText(getApplicationContext(), "Data saved.", Toast.LENGTH_LONG).show();
    }

    public void retrieveData() {
        sharedPreferences = getSharedPreferences("saveData", Context.MODE_PRIVATE);

        name = sharedPreferences.getString("name", null);
        message = sharedPreferences.getString("message", null);
        count = sharedPreferences.getInt("count", 0);
        isChecked = sharedPreferences.getBoolean("isChecked", false);

        userName.setText(name);
        userMessage.setText(message);
        counter.setText("" + count);
        remember.setChecked(isChecked);
    }
}

File: app/src/main/java/com/example/chap8/MainActivity.kt
package com.example.chap8

import android.content.Context
import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.Toast

class MainActivity : AppCompatActivity() {

    private lateinit var userName: EditText
    private lateinit var userMessage: EditText
    private lateinit var counter: Button
    private lateinit var remember: CheckBox

    var name: String? = null // This container is for the userName
    var message: String? = null // This container is for the userMessage
    var count: Int = 0 // This container is for the counter
    var isChecked: Boolean? = null // This container is for remember

    private lateinit var sharedPreferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        userName = findViewById(R.id.editTextText)
        userMessage = findViewById(R.id.editTextTextMultiLine)
        counter = findViewById(R.id.button)
        remember = findViewById(R.id.checkBox)

        counter.setOnClickListener {
            count++
            counter.text = count.toString()
        }
    }

    override fun onPause() {
        super.onPause()
        saveData()
    }

    override fun onResume() {
        super.onResume()
        retrieveData()
    }

    private fun saveData() {
        sharedPreferences = this.getSharedPreferences("saveData", Context.MODE_PRIVATE)
        name = userName.text.toString()
        message = userMessage.text.toString()
        isChecked = remember.isChecked

        val editor = sharedPreferences.edit()

        editor.putString("key name", name)
        editor.putString("key message", message)
        editor.putInt("key count", count)
        editor.putBoolean("key isChecked", isChecked!!)

        editor.apply() // Stores the data in the editor
        Toast.makeText(applicationContext, "Your data is saved.", Toast.LENGTH_LONG).show()
    }

    private fun retrieveData() {
        sharedPreferences = this.getSharedPreferences("saveData", Context.MODE_PRIVATE)

        name = sharedPreferences.getString("key name", null)
        message = sharedPreferences.getString("key message", null)
        count = sharedPreferences.getInt("key count", 0)
        isChecked = sharedPreferences.getBoolean("key isChecked", false)

        userName.setText(name)
        userMessage.setText(message)
        counter.text = count.toString()
        remember.isChecked = isChecked!!
    }
}